contributing.md

```diff
@@ -1,90 +1,207 @@
+To build the distribution files of BotD that can be used in a browser directly, run:
 
-One can create a new signal source in this client-side library, however, it needs to be also accepted and processed on the backend side properly. The recommended approach is to [file an issue in the repository](https://github.com/fingerprintjs/botd/issues) and discuss signal details with the maintainers.
+```bash
+yarn build
+```
+
+The files will be saved to the `dist` directory.
 
-#### How to add a new signal source to the client
+### How to test
 
-1. Create a new signal source in the `src/sources` directory.
-2. Register it in the `collector.ts` file, in the `collect()` function. The new signal and its enum number should be discussed with the maintainers of the repository.
+There are automated tests.
+They are run by [Jasmine](https://jasmine.github.io) in real browsers using [Karma](https://karma-runner.github.io).
+Unit test files are located right next to individual module files that they check.
+Integration tests are located in the `tests` directory.
 
-```ts
-import getNewSignal from './sources/getNewSignal'
-...
-[SignalName.MyNewSignal]: getNewSignal,
-...
+To run the tests in a browser on your machine, build the project and run:
+```bash
+yarn test:local --browsers ChromeHeadless
+# or to run in Firefox
+yarn test:local --browsers FirefoxHeadless
+# or to run in both
+yarn test:local
 ```
 
-### How to test
+To run the tests in browsers on [BrowserStack](https://www.browserstack.com), get a BrowserStack access key and run:
+```bash
+# For Linux, macOS and WSL (Linux on Windows)
+BROWSERSTACK_USERNAME=your-username BROWSERSTACK_ACCESS_KEY=your-key yarn test:browserstack
+```
+
+If you face `Error: spawn Unknown system error -86` on macOS, try installing Rosetta:
+```bash
+softwareupdate --install-rosetta
+```
+
+Alternatively, make a PR to this repository, the test will run on BrowserStack automatically.
+But the test won't run when the PR is made from a fork repository, in this case, a member will run the tests manually.
+
+BrowserStack sessions are unstable, so a session can fail for no reason;
+restart the testing when you see no clear errors related to the tests.
+If you run the test command multiple times in parallel, BrowserStack will lose access to the Karma server
+(for some reason), which will cause the tests to hang infinitely, so try to run a single test command at once.
+
+To check the distribution TypeScript declarations, build the project and run:
+
+```bash
+yarn check:dts
+```
+
+To check that the package is compatible with server-side rendering, build the project and run:
+
+```bash
+yarn check:ssr
+```
+
+### How to add a bot detector
```

Comment:
@xnerhu Please check this section for correctness and completeness. It should give outside contributors all information necessary for creating a new bot detector.

Reply:
Looks good, but I would also mention `src\utils\browser.ts`  for checking browser kind, engine ect

Reply:
Or type it in the comments, I'll add it

Reply:
I added it, feel free to modify it if needed

Reply:
I've reworder the paragraph, please take a look. I'll add the same paragraph to the FingerprintJS's contributing guide.